### ISSUE TITLE
Option to disable sorting JS arithmetic expressions

### ISSUE THREAD

**Issue by TikiTDO** [snowcoders/sortier#268]
Sortier re-organizes arithmetic expressions which may be ordered in a particular way based on the formulae being applied. The resulting code can lose a lot of implicit meaning without offering any particular benefits. An option to disable this particular sorting behavior would be really helpful.

# Language

Javascript

# Sample Input

Trying a sort on a file that calculates 5 times the area of a circle

```js
const example = 2 * pi * distance * distance * 5
```

# Expected output

Expecting no change

```js
const example = 2 * pi * distance * distance * 5
```

# Actual output

Actual result is a blob made up of all the components. This is particularly an issue with longer, more complex calculations which can become practically illegible after sorting.

```js
const example = 2 * 5 * distance * distance * pi
```

**Comment by k2snowman69** [comment]
Hmm cool idea @TikiTDO and a great use case! I can see a few ways this could be implemented:

### Disabling it for all
This would be in the `src\language-js\sortExpression\index.ts` and basically on the specific AST element of an arithmetic expression (I don't remember what it is off the top of my head) just returning `fileContents` in `sortExpression`. The option in the sortier config would be down the path `options.js.sortExpression.disable<InsertExpressionName>`. In this situation, we need to be careful not to disable it for union types and other general expression type situations.

### Disable by comment
Basically, if a AST tree element has a `// sortier-ignore` or something similar as a comment above the element to be sorted, we should just ignore it. I'm not sure how this would interact with combination scenarios such as this arithmetic situation + sortClassContents for example. I think that would be something that would have to be well investigated and tested.

I think the first would likely be easier to implement because it's far more scoped but the second maybe more useful. Did you have any other ideas of how you would like this to work?

**Comment by TikiTDO** [comment]
Does sortier support comment config clauses? I didn't realize that was an option. That would go a long way when it comes to those one-offs that should absolutely not be sorted. Just use it to disable sorting for all nodes and sub-nodes that occur on the next line, though it would be of limited utility for code bases where these things are all over the place.

The idea of disabling by AST type is feels right though, it would certainly offer a lot of flexibility. On the other hand it would probably means more support requests, since, you know, a lot of flexibility. Going in that direction I'd recommend expanding the README with examples for people that are not familiar with compilers. 

As a middle ground it may be possible to do a set of special handling options which could be used to enable / disable special behavior, and a separate file to configure said behaviors. This could even be extended to offer users the ability to add their own sorting options on a per-project basis.

I'm pretty busy til the weekend, but when I free up I could poke around the code to see if there are clean ways to extend the functionality.

**Comment by k2snowman69** [comment]
Sortier doesn't right now, I was just suggesting ideas and any direction we go with this would require updating the readme.

I'd rather avoid disabling by AST type in general; when I said `options.js.sortExpression.disable<InsertExpressionName>` I meant specifically whichever is for the arithmetic expression, not generically for any expression. I think the generic route would be a bit too flexible for where the tool is right now.

I can also see people wanting to sort certain arithmetic expressions but not others which is why I brought up the disable by comment instead as it would allow for that flexibility but the core code of asking "should I sort this" could be located and likely implemented in a single location (src/utilities/sort-utils.ts for example) without having to impact the rest of the code base.

Feel free to poke around, I'm pretty busy myself over the next week but I'm hoping to enable sorting of comments to the right of a line (not just left and above) which would cause a v3 bump. Overall the rest of the code base isn't going to change much so you shouldn't run into merge conflicts :-)

**Comment by k2snowman69** [comment]
Did you get time to look into this yet? If not I'm going to look into the `// sortier-ignore` option this weekend most likely.

**Comment by TikiTDO** [comment]
Ah sorry, things blew up a bit. Gonna be fire-fighting for a while.

**Comment by k2snowman69** [comment]
I created https://github.com/snowcoders/sortier/tree/k2snowman69/sortierignore as a first run through. If `sortier-ignore` is found in a comment above the node, sortier will skip that node's entire tree. 

I've created a test case to help us think through a few scenarios here... the import declaration scenario is a bit interesting. May take a bit of playing around to get this to feel right before we merge it in.

**Comment by k2snowman69** [comment]
I had some thoughts on this and played a bit more with it. I've updated the branch above with two new comments, `sortier-ignore-line` and `sortier-ignore-node`. Effectively, line ignores any nodes that begin and end in the next line where node will ignore the outer most node and anything within it. There are definite edge cases I'm not hitting here (sorting of switch cases, sorting of import statements, etc).

I'm also tempted to delete the node option all together for now as personally I think it's just a bit too powerful and has far more scenarios.

The only thing preventing both of them is the failing test where the `sortier-ignore-*` should move with the code even though it's technically a comment for the context group (which we don't move generally but in this case I think we should).

**Comment by k2snowman69** [comment]
3.2.0
    - Added sortier-ignore-next-line support which disables sorting on all AST nodes that start and end on the next line
    - Added sortier-ignore-nodes support which disables sorting on all AST nodes that start on the next line
